File: app/src/main/java/com/example/islandlyrics/LyricRepository.java
package com.example.islandlyrics;

import androidx.lifecycle.MutableLiveData;

/**
 * Singleton repository to hold lyric state.
 */
public class LyricRepository {
    private static LyricRepository sInstance;

    private final MutableLiveData<Boolean> isPlaying = new MutableLiveData<>(false);
    private final MutableLiveData<String> currentLyric = new MutableLiveData<>("Thinking...");
    private final MutableLiveData<String> sourceAppName = new MutableLiveData<>("No Source");
    private final MutableLiveData<String> songTitle = new MutableLiveData<>("Unknown Title");
    private final MutableLiveData<String> artistName = new MutableLiveData<>("Unknown Artist");
    
    // New Atomic Metadata Container
    public static class MediaInfo {
        public final String title;
        public final String artist;
        public final String packageName;
        public final long duration;
        
        public MediaInfo(String title, String artist, String packageName, long duration) {
            this.title = title;
            this.artist = artist;
            this.packageName = packageName;
            this.duration = duration;
        }
    }
    
    // New Atomic Lyric Container
    public static class LyricInfo {
        public final String lyric;
        public final String sourceApp;
        
        public LyricInfo(String lyric, String sourceApp) {
            this.lyric = lyric;
            this.sourceApp = sourceApp;
        }
    }
    
    private final MutableLiveData<MediaInfo> liveMetadata = new MutableLiveData<>();
    private final MutableLiveData<LyricInfo> liveLyric = new MutableLiveData<>();

    private LyricRepository() {}

    public static synchronized LyricRepository getInstance() {
        if (sInstance == null) {
            sInstance = new LyricRepository();
        }
        return sInstance;
    }

    // Getters for LiveData
    public MutableLiveData<Boolean> getIsPlaying() { return isPlaying; }
    public MutableLiveData<String> getCurrentLyric() { return currentLyric; }
    public MutableLiveData<String> getSourceAppName() { return sourceAppName; }
    public MutableLiveData<String> getSongTitle() { return songTitle; }
    public MutableLiveData<String> getArtistName() { return artistName; }
    public MutableLiveData<MediaInfo> getLiveMetadata() { return liveMetadata; }
    public MutableLiveData<LyricInfo> getLiveLyric() { return liveLyric; }

    // Update methods
    public void updatePlaybackStatus(boolean playing) {
        if (isPlaying.getValue() != playing) {
            isPlaying.postValue(playing);
        }
    }

    public void updateLyric(String lyric, String app) {
        if (lyric != null) currentLyric.postValue(lyric);
        if (app != null) sourceAppName.postValue(app);
        
        // Atomic Update
        if (lyric != null && app != null) {
            liveLyric.postValue(new LyricInfo(lyric, app));
        }
        
        // This method NO LONGER updates title/artist to prevent overwriting metadata from MediaMonitor
        
        // Implicitly playing if we get data
        updatePlaybackStatus(true);
    }

    public void updateMediaMetadata(String title, String artist, String packageName, long duration) {
        AppLogger.getInstance().log("Repo", "Posting metadata for: " + packageName);
        // Atomic update
        liveMetadata.postValue(new MediaInfo(title, artist, packageName, duration));
    }
}

File: app/src/main/java/com/example/islandlyrics/LyricRepository.kt
package com.example.islandlyrics

import androidx.lifecycle.MutableLiveData

/**
 * Singleton repository to hold lyric state.
 */
class LyricRepository private constructor() {

    // Playback state
    val isPlaying = MutableLiveData(false)

    // Atomic Metadata Container
    data class MediaInfo(
        val title: String,
        val artist: String,
        val packageName: String,
        val duration: Long
    )

    // Atomic Lyric Container
    data class LyricInfo(
        val lyric: String,
        val sourceApp: String
    )

    // Playback Progress Container
    data class PlaybackProgress(
        val position: Long,  // in milliseconds
        val duration: Long   // in milliseconds
    )

    // Modern atomic LiveData (single source of truth)
    val liveMetadata = MutableLiveData<MediaInfo?>()
    val liveLyric = MutableLiveData<LyricInfo?>()
    val liveProgress = MutableLiveData<PlaybackProgress?>()

    // Track previous song to detect changes
    private var lastTrackId: String? = null

    // Update methods
    fun updatePlaybackStatus(playing: Boolean) {
        if (isPlaying.value != playing) {
            isPlaying.postValue(playing)
        }
    }

    fun updateLyric(lyric: String?, app: String?) {
        // Atomic Update: Only post if both lyric and app are non-null
        if (lyric != null && app != null) {
            liveLyric.postValue(LyricInfo(lyric, app))
        }

        // Implicitly playing if we get lyric data
        updatePlaybackStatus(true)
    }

    fun updateMediaMetadata(title: String, artist: String, packageName: String, duration: Long) {
        // Simple atomic update - no song change detection, no lyric clearing
        // Capsule lifecycle is now controlled by playback state, not metadata changes
        AppLogger.getInstance().log("Repo", "📝 Metadata: $title - $artist [$packageName]")
        liveMetadata.postValue(MediaInfo(title, artist, packageName, duration))
    }

    fun updateProgress(position: Long, duration: Long) {
        liveProgress.postValue(PlaybackProgress(position, duration))
    }

    companion object {
        private var instance: LyricRepository? = null

        @Synchronized
        fun getInstance(): LyricRepository {
            if (instance == null) {
                instance = LyricRepository()
            }
            return instance!!
        }
    }
}
